plugin: add queue update validation

cmoussa1 opened the pull request:
#### Background

As noted in #377, Flux can now handle job updates to pending jobs, particularly with queues. Since flux-accounting's priority plugin is involved in queue validation (i.e making sure that the user/bank has access to the queue they want to submit their job to), it should be expanded to handle job updates.

---

This PR adds a new callback to the priority plugin for `job.update`, which will unpack user, bank, and queue information and assign the new priority associated with the updated queue to the `bank_info` struct of a job so that when the job re-enters `job.state.priority`, a new overall job priority can be calculated.

A new helper function is added to the plugin called `get_bank_info ()`, which accepts a userid and an optional bank value (if one is passed during job submission) that performs a lookup in the plugin's internal map to retrieve the necessary information to associate with the submitted job. It returns a pair: a return code as well as the iterator to the user/bank map that is associated with the job. The `job.update` callback makes use of this to retrieve the user/bank information, and then retrieves the new priority associated with the updated queue and assigns it to the `bank_info` struct associated with a job, which could affect its priority.

This new helper function can most likely be adapted in other places in the plugin code, but I think its best to probably leave it for a future PR to refactor some of the callbacks to make use of it.

Finally, a new set of tests are added to the testsuite which test updating the queue for a job in which the user/bank **does** have access to and updating the queue for a job in which the user/bank **does not** have access to. In the former, the job is updated and the priority is increased, and in the latter, an error is raised when the job re-enters `job.validate` and the priority remains the same as it was before the attempted update. Note that these tests are commented out until flux-framework/flux-core#5424 has landed; otherwise, I think these tests will fail in the meantime.

cmoussa1 committed:
plugin: add function to retrieve user/bank info

Create a helper function, get_bank_info (), that accepts a userid and a
bank parameter and performs a lookup in the plugin's internal users map.
It returns a pair: a return code as well as the iterator to the
user/bank map that contains the information to be associated with each
job's bank_info struct, such as their job limits and permissible
queues.

cmoussa1 committed:
plugin: add job-update callback

Problem: Flux can now accept job updates, but the multi-factor priority
plugin does not have a callback to look at these updates, particularly
for attributes that it is involved in validating.

Add a job-update callback to the plugin. Unpack the userid, bank, and
queue attributes from jobspec and assign the priority associated with
the new queue if one is passed in to the bank_info struct that is
associated with the job.

grondo commented on src/plugins/mf_priority.cpp:
```diff
@@ -45,6 +45,21 @@ extern "C" {
+// BANK_SUCCESS: we found an entry for the passed-in user/bank
+// INVALID_BANK: the user specified a bank they don't belong to
+// NO_DEFAULT_BANK: the user does not have a default bank in the plugin map
+enum bank_info_codes {
+    USER_NOT_FOUND,
+    BANK_SUCCESS,
+    INVALID_BANK,
+    NO_DEFAULT_BANK
+};
```
Suggestion: it is sometimes good practice to name enumerations with a common prefix, e.g. if it works here prefix all these values with `BANK_`: `BANK_USER_NOT_FOUND`, `BANK_NO_DEFAULT`, `BANK_INVALID` etc. This just helps when reading and writing code.

Also, just another suggestion, but is more typical to see the first enumeration value (i.e. `0`) be SUCCESS. Since that is more common practice, I'd suggest moving `BANK_SUCCES` to the first position of the enumeration.

cmoussa1 replied:
Thanks for these suggestions! I was unaware of the practices here (I think it's my first time using enumeration values? can't remember if i ever did this), so I'll make these changes to use a common `BANK` prefix as well as using the SUCCESS value as the first position. 👍

grondo commented on t/t1029-mf-priority-update-queue.t:
```diff
@@ -0,0 +1,87 @@
+# 	flux job eventlog $jobid1 > eventlog.out &&
+# 	grep "attributes.system.queue=\"silver\"" eventlog.out
+# '
+
```
Is there a way to check that the update had the intended effect, i.e. check for a new priority (or maybe query the mf_priority plugin and ensure the queue factor has been updated?

cmoussa1 replied:
Yes there is. I forgot to add those to the official test; was just printing out the priority values before, but I think I can check priority values with `flux job wait-event -f json $jobid priority | jq '.context.priority'`. I'll add these to the testsuite. Thanks!

grondo reviewed (commented):
Looks good! Just a quick first pass with a couple suggestions, but otherwise I think this looks good!

cmoussa1 committed:
t: add tests for updating queue

Problem: flux-accounting has no tests that update the queue of a pending
job and ensures that the priority plugin can handle cases where the job
should be updated and where the job should not be updated because the
queue cannot be validated for the user/bank's job.

Add some tests.

grondo commented:
flux-framework/flux-core#5424 has been merged, so tests could be enabled here now.

cmoussa1 commented:
OK, I've uncommented out the tests that were dependent on flux-framework/flux-core#5424 and rebased; I'll mark it as ready for some initial review.

grondo commented on src/plugins/mf_priority.cpp:
```diff
@@ -1074,6 +1074,71 @@ static int run_cb (flux_plugin_t *p,
+    char *queue = NULL;
+    struct bank_info *b;
+
+    flux_t *h = flux_jobtap_get_flux (p);
```
I could have missed it, but it appears `h` is assigned here and never used?

grondo commented on t/Makefile.am:
```diff
@@ -31,6 +31,7 @@ TESTSCRIPTS = \
 	t1027-mf-priority-issue376.t \
 	t1028-mf-priority-issue385.t \
 	t1029-mf-priority-default-bank.t \
+	t1030-mf-priority-update-queue.t \
```
be sure to remove the `note:` from the commit message here.

grondo commented on t/t1030-mf-priority-update-queue.t:
```diff
@@ -0,0 +1,92 @@
+
+test_expect_success 'updating a job using a queue the user does not belong to fails' '
+	test_must_fail flux update $jobid1 queue=gold > unavail_queue.out 2>&1 &&
+	test_debug "unavail_queue.out" &&
```
This should be `test_debug "cat unavail_queue.out"`

grondo reviewed (approved):
Looks good! Just a couple minor comments.

cmoussa1 commented:
Thanks for the review @grondo! I've pushed up the fixes you mentioned in your comments. Since you've approved, I'll set MWP on this. Thanks again!!
